src/StringTools/StringTools.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+#if !NET35
+using Microsoft.Extensions.ObjectPool;
+#endif
+
+namespace Microsoft.StringTools
+{
+    /// <summary>
+    /// A callback type to register with the library. The library will invoke this callback in registration order until
+    /// true is returned. If no callback returns true, the library will fall back to its default interning logic.
+    /// </summary>
+    /// <param name="internableString">The string to be interned.</param>
+    /// <param name="result">The interned string. Set to null and return true if the string should not be interned.</param>
+    /// <returns>True if the <paramref name="internableString" /> was handled by the callback.</returns>
+    public delegate bool TryInternStringDelegate(ref InternableString internableString, out string result);
+
+    public static class Strings
+
+    {
+#if !NET35
+        /// <summary>
+        /// IPooledObjectPolicy used by <cref see="s_stringBuilderPool"/>.
+        /// </summary>
+        private class PooledObjectPolicy : IPooledObjectPolicy<SpanBasedStringBuilder>
+        {
+            /// <summary>
+            /// No need to retain excessively long builders forever.
+            /// </summary>
+            private const int MAX_RETAINED_BUILDER_CAPACITY = 1000;
+
+            /// <summary>
+            /// Creates a new SpanBasedStringBuilder with the default capacity.
+            /// </summary>
+            public SpanBasedStringBuilder Create()
+            {
+                return new SpanBasedStringBuilder();
+            }
+
+            /// <summary>
+            /// Returns a builder to the pool unless it's excessively long.
+            /// </summary>
+            public bool Return(SpanBasedStringBuilder stringBuilder)
+            {
+                if (stringBuilder.Capacity <= MAX_RETAINED_BUILDER_CAPACITY)
+                {
+                    stringBuilder.Clear();
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// A pool of SpanBasedStringBuilders as we don't want to be allocating every time a new one is requested.
+        /// </summary>
+        private static DefaultObjectPool<SpanBasedStringBuilder> s_stringBuilderPool =
+            new DefaultObjectPool<SpanBasedStringBuilder>(new PooledObjectPolicy(), Environment.ProcessorCount);
+#endif
+
+        /// <summary>
+        /// An array of callbacks to be called for each string being interned.
+        /// </summary>
+        private static volatile TryInternStringDelegate[] s_internStringCallbacks = new TryInternStringDelegate[0];
```

Comment:
Would you mind explaining every part of this?

Also, nit:
```suggestion
        private static volatile TryInternStringDelegate[] s_internStringCallbacks = Array.Empty<TryInterStringDelegate>();
```

Reply:
Actually, should this be an array? I didn't see anywhere that you used it without looking at every element, and you sometimes add to the end. Maybe a LinkedList? Arrays have less overhead, but a LinkedList might be faster...

Reply:
I wanted a data structure that can be read in a lock-free fashion. Since writes are expected to be super rare, I went with the simplest type - an immutable array. Readers get a reference to the array and read it lock-free because no writer is modifying the array once published into the static. Only writers have to synchronize between them so they don't step on each other toes. Done with a lock here, could be a loop with an interlocked exchange as well. Also note the `volatile` modifier which ensures that the contents of the array is visible to other threads when a writer does `s_internStringCallbacks =`.

But maybe all this is moot if, as Dan suggested, we don't need this callback mechanism at all. I am going to measure this shortly.